File: Sem3/Advanced Methods of Programming/MyInterpreter - Java Console Application/project/src/domain/ProgramState.java
package domain;


import controller.ControllerException;
import domain.adt.*;
import domain.statement.Heap.HeapAddressBuilder;
import domain.statement.StatementException;
import domain.statement.StatementInterface;


public class ProgramState
{
    private MyStackInterface<StatementInterface> exeStack;
    private MyDictionaryInterface<String, Integer> symbolTable;
    private MyListInterface<Integer> out;
    private StatementInterface originalProgram; //optional
    private MyDictionaryInterface<Integer, MyPair> fileTable;
    private MyDictionaryInterface<Integer, Integer> heapTable;
    private HeapAddressBuilder addressBuilder = new HeapAddressBuilder();
    private int threadID;


    public ProgramState(MyStackInterface<StatementInterface> exeStack,
                        MyDictionaryInterface<String, Integer> symbolTable,
                        MyListInterface<Integer> out,
                        MyDictionaryInterface<Integer, MyPair> fileTable,
                        MyDictionaryInterface<Integer, Integer> heapTable,
                        Integer threadID,
                        StatementInterface program)

    {
        this.exeStack = exeStack;
        this.symbolTable = symbolTable;
        this.out = out;
        //this.originalProgram = deepCopy(program);
        this.fileTable = fileTable;
        this.heapTable = heapTable;
        this.threadID = threadID;
        exeStack.push(program);
    }


    public MyStackInterface<StatementInterface> getExeStack()
    {
        return exeStack;
    }


    public MyDictionaryInterface<String, Integer> getSymbolTable()
    {
        return symbolTable;
    }


    public void setSymbolTable(MyDictionaryInterface<String, Integer> symbolTable)
    {
        this.symbolTable = symbolTable;
    }


    public MyListInterface<Integer> getOutput()
    {
        return out;
    }


    public StatementInterface getOriginalProgram()
    {
        return originalProgram;
    }


    public MyDictionaryInterface<Integer, MyPair> getFileTable()
    {
        return fileTable;
    }


    public MyDictionaryInterface<Integer, Integer> getHeapTable()
    {
        return heapTable;
    }


    @Override
    public String toString()
    {
        return "threadID: " + threadID + "\n" +
                "ExeStack:\n" + exeStack.toString() +
                "SymbolTable:\n" + symbolTable.toString() +
                "Print output:\n" + out.toString() +
                "FileTable:\n" + fileTable.toString() +
                "HeapTable:\n" + heapTable.toString();
    }


    public Integer getNewAddress()
    {
        return addressBuilder.getFreeAddress();
    }


    public void setExeStack(MyStackInterface<StatementInterface> exeStack)
    {
        this.exeStack = exeStack;
    }


    public boolean isNotCompleted()
    {
        return !exeStack.isEmpty();
    }


    public ProgramState oneStep() throws StatementException
    {
        try
        {
            StatementInterface currentStatement = exeStack.pop();
            return currentStatement.execute(this);
        }
        catch (AdtException exception)
        {
            throw new StatementException(exception.getMessage());
        }
    }
}

File: Sem3/Advanced Methods of Programming/MyInterpreter - CSharp Console Application/MyInterpreter CSharp/domain/ProgramState.cs
using System.IO;
using MyInterpreter_CSharp.domain.adt;
using MyInterpreter_CSharp.domain.statement;

namespace MyInterpreter_CSharp.domain
{
    public class ProgramState
    {
        private IMyStack<IStatement> _exeStack;
        private IMyDictionary<string, int> _symbolTable;
        private IMyList<int> _output;
        private IMyDictionary<int, MyPair<string, TextReader>> _fileTable;

        public ProgramState(
            IMyStack<IStatement> exeStack,
            IMyDictionary<string, int> symbolTable,
            IMyList<int> output,
            IMyDictionary<int, MyPair<string, TextReader>> fileTable,
            IStatement program)
        {
            _exeStack = exeStack;
            _symbolTable = symbolTable;
            _output = output;
            _fileTable = fileTable;
            exeStack.Push(program);
        }

        public IMyStack<IStatement> ExeStack => _exeStack;

        public IMyDictionary<string, int> SymbolTable => _symbolTable;

        public IMyList<int> Output => _output;

        public IMyDictionary<int, MyPair<string, TextReader>> FileTable => _fileTable;

        public override string ToString()
        {
            return "ExeStack:\n" + _exeStack.ToString() +
                   "SymbolTable:\n" + _symbolTable.ToString() +
                   "FileTable:\n" + _fileTable.ToString() +
                   "Print output:\n" + _output.ToString();
        }
    }
}